File: src/main/java/club/maxstats/weave/remapping/MinecraftRemapper.java
package club.maxstats.weave.remapping;

import club.maxstats.weave.configuration.MinecraftVersion;
import lombok.experimental.UtilityClass;
import org.objectweb.asm.commons.Remapper;
import org.objectweb.asm.commons.SimpleRemapper;

import java.io.*;
import java.util.HashMap;
import java.util.Map;

@UtilityClass
public class MinecraftRemapper {

    public Remapper create(MinecraftVersion version) throws IOException {
        try (InputStream mappings = version.getMappingStream()) {
            return new SimpleRemapper(parseMappings(mappings));
        }
    }

    /**
     * Parses the mappings from the specified stream.
     *
     * @param mappingsStream The stream from which to read the mappings
     * @return The mappings according to their corresponding version.
     */
    private Map<String, String> parseMappings(InputStream mappingsStream) throws IOException {
        Map<String, String> mappings = new HashMap<>();
        BufferedReader           rdr = new BufferedReader(new InputStreamReader(mappingsStream));

        while (rdr.ready()) {
            String line = rdr.readLine();
            if (line.isEmpty()) continue;

            String[] split = line.substring(4).split(" ");
            switch (line.substring(0, 4)) {
                case "CL: ": {
                    mappings.put(split[0], split[1]);
                    break;
                }
                case "MD: ": {
                    int    i      = split[0].lastIndexOf('/');
                    String clazz  = split[0].substring(0, i);
                    String method = split[0].substring(i + 1);

                    mappings.put(String.format("%s.%s%s", clazz, method, split[1]), split[2].substring(split[2].lastIndexOf('/') + 1));
                    break;
                }
                case "FD: ": {
                    int    i     = split[0].lastIndexOf('/');
                    String clazz = split[0].substring(0, i);
                    String field = split[0].substring(i + 1);

                    mappings.put(String.format("%s.%s", clazz, field), split[2].substring(split[2].lastIndexOf('/') + 1));
                    break;
                }
            }
        }

        return mappings;
    }

}

File: src/main/java/club/maxstats/weave/configuration/MinecraftRemapper.kt
package club.maxstats.weave.configuration

import org.objectweb.asm.commons.SimpleRemapper
import java.io.InputStream

fun createMinecraftRemapper(version: MinecraftVersion) =
    version.mappingStream.use { SimpleRemapper(it.parseMappings()) }

/**
 * Parses the mappings from the specified stream.
 *
 * @receiver The stream from which to read the mappings
 * @return The mappings according to their corresponding version.
 */
private fun InputStream.parseMappings() = bufferedReader().useLines { seq ->
    seq.filter { it.isNotEmpty() }.associate { line ->
        val split = line.drop(4).split(" ")
        when (line.take(4)) {
            "CL: " -> split[0] to split[1]

            "MD: " -> {
                val clazz  = split[0].substringBeforeLast('/')
                val method = split[0].substringAfterLast('/')

                "$clazz.$method${split[1]}" to split[2].substringAfterLast('/')
            }

            "FD: " -> {
                val clazz = split[0].substringBeforeLast('/')
                val field = split[0].substringAfterLast('/')

                "$clazz.$field" to split[2].substringAfterLast('/')
            }

            else -> error("Invalid mapping sequence")
        }
    }
}
